File: tests/regression/400_RUNTIME_FEATURES/430_RUNTIME/430_014_interpreter_benchmark/honest_bench.js
class PRNG {
    constructor(seed) { this.state = BigInt(seed); }
    next(max) {
        this.state = (this.state * 6364136223846793005n + 1442695040888963407n) & 0xFFFFFFFFFFFFFFFFn;
        return Number((this.state >> 33n) % BigInt(max));
    }
}

function add_handler(a, b) { return a + b; }
function mul_handler(a, b) { return a * b; }
function sub_handler(a, b) { return a - b; }
function div_handler(a, b) { return b !== 0 ? Math.trunc(a / b) : 0; }

const handlers = { add: add_handler, mul: mul_handler, sub: sub_handler, div: div_handler };
const events = ["add", "mul", "sub", "div"];

function dispatch(eventName, a, b) {
    const handler = handlers[eventName];
    if (handler) return handler(a, b);
    throw new Error("Unknown event");
}

const ITERATIONS = 10_000_000;
const prng = new PRNG(12345);

const start = process.hrtime.bigint();

let sum = 0;
for (let i = 0; i < ITERATIONS; i++) {
    const eventIdx = prng.next(4);
    const a = prng.next(100) + 1;
    const b = prng.next(100) + 1;
    sum += dispatch(events[eventIdx], a, b);
}

const end = process.hrtime.bigint();
const elapsed_ms = Number(end - start) / 1_000_000;
const ops_per_sec = ITERATIONS / (Number(end - start) / 1_000_000_000);

console.log("HONEST Node: " + elapsed_ms.toFixed(2) + "ms, " + Math.round(ops_per_sec) + " ops/sec, sum=" + sum);

File: tests/regression/400_RUNTIME_FEATURES/430_RUNTIME/430_014_interpreter_benchmark/honest_bench.py
import time

class PRNG:
    def __init__(self, seed):
        self.state = seed
    def next(self, max_val):
        self.state = (self.state * 6364136223846793005 + 1442695040888963407) & 0xFFFFFFFFFFFFFFFF
        return (self.state >> 33) % max_val

def add_handler(a, b): return a + b
def mul_handler(a, b): return a * b
def sub_handler(a, b): return a - b
def div_handler(a, b): return a // b if b != 0 else 0

handlers = {"add": add_handler, "mul": mul_handler, "sub": sub_handler, "div": div_handler}
events = ["add", "mul", "sub", "div"]

def dispatch(event_name, a, b):
    return handlers[event_name](a, b)

ITERATIONS = 10_000_000
prng = PRNG(12345)

start = time.perf_counter_ns()

total = 0
for i in range(ITERATIONS):
    event_idx = prng.next(4)
    a = prng.next(100) + 1
    b = prng.next(100) + 1
    total += dispatch(events[event_idx], a, b)

end = time.perf_counter_ns()
elapsed_ms = (end - start) / 1_000_000
ops_per_sec = ITERATIONS / ((end - start) / 1_000_000_000)

print(f"HONEST Python: {elapsed_ms:.2f}ms, {int(ops_per_sec)} ops/sec, sum={total}")
